types/errors/errors.go

```diff
@@ -123,6 +123,14 @@ var (
 	// ErrUnpackAny defines an error when unpacking a protobuf message from Any fails.
 	ErrUnpackAny = Register(RootCodespace, 34, "failed unpacking protobuf message from Any")
 
+	// ErrLogic defines an internal logic error, e.g. an invariant or assertion
+	// that is violated. It is a programmer error, not a user-facing error.
+	ErrLogic = Register(RootCodespace, 35, "internal logic error")
+
+	// ErrConflict defines a conflict error, e.g. when two goroutines try to access
+	// the same resource and one of them fails.
+	ErrConflict = Register(RootCodespace, 36, "conflict")
```

Comment:
These are the current scenarios that trigger `ErrConflict`:

* The BaseApp tries to start a snapshot/prune operation while a different snapshot/prune operation is in progress (e.g. when creating a snapshot takes longer than the snapshot interval).

* A snapshot is created for a height that already has a snapshot.

* A snapshot is attempted created or deleted while it is being created.

Conflict is the usual nomenclature for these sorts of concurrency issues. We could call it e.g. `ErrRace`, but I'm not sure if that's any clearer, and it's not always a race condition (e.g. block replay can cause the application to attempt to create a snapshot that has already been created, which isn't a race, simply non-idempotency).

Reply:
Totally get that, but when I see "conflict", I don't automatically get that context. I don't think this is expressive enough to infer that. Unless you already know this, "conflict" could mean anything.

Reply:
Ah - since this is always wrapped with an error message, the context sort of follows from that. But maybe I'm using this error convention wrong. Would you prefer having individual error types for these cases? I'm not really sure what the alternative would be.

Reply:
No, you're right. I took a look at all the uses of `ErrConflict` and the context is sufficient. I would say we could rename this to `SnapshotConflict`, but we may end up having to use a similar error for other parts of the codebase in the future, so `ErrConflict` is OK 👍 

resolving...

Reply:
I prefer `ErrConflict`. The context about Snapshot can be added in a wrapper.